disttask/loaddata/job.go

```diff
@@ -0,0 +1,151 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package loaddata
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/disttask/framework/proto"
+	"github.com/pingcap/tidb/disttask/framework/storage"
+	"github.com/pingcap/tidb/executor/importer"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+var (
+	distLoadDataConcurrency = 16
+	checkTaskFinishInterval = 300 * time.Millisecond
+)
+
+// DistImporter is a JobImporter for distributed load data.
+type DistImporter struct {
+	*importer.JobImportParam
+	plan *importer.Plan
+}
+
+var _ importer.JobImporter = &DistImporter{}
+
+// NewDistImporter creates a new DistImporter.
+func NewDistImporter(param *importer.JobImportParam, plan *importer.Plan) (*DistImporter, error) {
+	return &DistImporter{
+		JobImportParam: param,
+		plan:           plan,
+	}, nil
+}
+
+// Param implements JobImporter.Param.
+func (ti *DistImporter) Param() *importer.JobImportParam {
+	return ti.JobImportParam
+}
+
+// Import implements JobImporter.Import.
+func (ti *DistImporter) Import() {
+	logutil.BgLogger().Info("start distribute load data", zap.Int64("jobID", ti.Job.ID))
+	ti.Group.Go(func() error {
+		defer close(ti.Done)
+		return ti.doImport(ti.GroupCtx)
+	})
+}
+
+// Result implements JobImporter.Result.
+func (*DistImporter) Result() importer.JobImportResult {
+	return importer.JobImportResult{}
+}
+
+// Close implements the io.Closer interface.
+func (*DistImporter) Close() error {
+	return nil
+}
+
+func buildDistTask(plan *importer.Plan, jobID int64) TaskMeta {
+	return TaskMeta{
+		Plan:  *plan,
+		JobID: jobID,
+	}
+}
+
+// submitGlobalTaskAndRun submits a global task and returns a channel that will be closed when the task is done.
+// TODO: move to handle, see https://github.com/pingcap/tidb/pull/43066
+func submitGlobalTaskAndRun(ctx context.Context, taskKey, taskType string, concurrency int, taskMeta []byte) error {
+	globalTaskManager, err := storage.GetTaskManager()
+	if err != nil {
+		return err
+	}
+	globalTask, err := globalTaskManager.GetGlobalTaskByKey(taskKey)
+	if err != nil {
+		return err
+	}
+
+	if globalTask == nil {
+		taskID, err := globalTaskManager.AddNewGlobalTask(taskKey, taskType, concurrency, taskMeta)
+		if err != nil {
+			return err
+		}
+
+		globalTask, err = globalTaskManager.GetGlobalTaskByID(taskID)
+		if err != nil {
+			return err
+		}
+
+		if globalTask == nil {
+			return errors.Errorf("cannot find global task with ID %d", taskID)
+		}
+	}
+
+	ticker := time.NewTicker(checkTaskFinishInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
```

Comment:
what's the root cause of context cancel, can we return nil here to avoid filtering context error later?

Reply:
add comment in https://github.com/pingcap/tidb/pull/43066